File: 0271-encode-and-decode-strings/0271-encode-and-decode-strings.js
/**
 * Encodes a list of strings to a single string.
 *
 * @param {string[]} strs
 * @return {string}
 */
var encode = function(strs) {
    const res = []
    for(const str of strs){
        res.push(str.length + "#" + str)
    }
    return res.join("")    
};

/**
 * Decodes a single string to a list of strings.
 *
 * @param {string} s
 * @return {string[]}
 */
var decode = function(s) {
    const res = []
    let i = 0
    while(i < s.length){
        let j = i
        while(s[j] != "#"){ // get length number
            j++
        } // j is #
        const length = Number(s.slice(i, j))
        const word = s.slice(j+1, j+1+length)
        res.push(word)
        i = j+1+length //now i is beginning of the next word
    }

    return res
};

/**
 * Your functions will be called as such:
 * decode(encode(strs));
 */

 // str length # str

File: 0271-encode-and-decode-strings/0271-encode-and-decode-strings.py
class Codec:
    def encode(self, strs: List[str]) -> str:
        """Encodes a list of strings to a single string.
        """
        res = []
        for s in strs:
            res.append(f"{len(s)}#{s}")
        return "".join(res)
        

    def decode(self, s: str) -> List[str]:
        """Decodes a single string to a list of strings.
        """
        res = []
        start = 0
        end = 0
        while end < len(s):
            if s[end] != "#":
                end += 1
            else:
                length = int(s[start:end])
                word = s[end+1:end+1+length]
                res.append(word)
                start = end+1+length
                end = start
        
        return res



        


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(strs))